README.md
# localModel: Local Explanations of Machine Learning Models for Tabular Data.

[![CRAN_Status_Badge](http://www.r-pkg.org/badges/version/localModel)](https://cran.r-project.org/package=localModel)
[![Travis-CI Build Status](https://travis-ci.org/ModelOriented/localModel.svg?branch=master)](https://travis-ci.org/ModelOriented/localModel)
[![Coverage Status](https://img.shields.io/codecov/c/github/ModelOriented/localModel/master.svg)](https://codecov.io/github/ModelOriented/localModel?branch=master)


`localModel` is a successor to the `live` package. It implements a variant of LIME method for explaining single predictions of black box machine learning models for tabular data.
Interpretable features are created based on [Ceteris Paribus](https://github.com/ModelOriented/ceterisParibus2) plots.
Details of the methodology are described in the vignette.
`localModel` is currently undergoing rapid changes, including bug fixes, for a stable solution please see the `live` package. 

To get started, install the package from CRAN:

```
install.packages('localModel')
```


The development version can be installe from GitHub by using the code below. 
Please do not use the `devtools` package, since it is affected [a bug](https://github.com/r-lib/devtools/issues/1900) which makes `localModel` installation impossible.
This issue was solved in the Github version of `remotes`.

```
devtools::install_github("r-lib/remotes")
remotes::install_github("ModelOriented/localModel")
```

To get help, see examples and details of the methodology, please refer to package website and vignettes.

## Acknowledgments 

Work on this package is financially supported by the NCN Opus grant 2017/27/B/ST6/01307.


NEWS.md
# localModel 0.3.12 

* Fixes problems in tests.

# localModel 0.3.11

* Changes dependencies: `DALEX2` to `DALEX`, `ceterisParibus2` to `ingredients`.

# localModel 0.3.10

* Vignettes updated. In particular, seed is now set before using randomForest.
* Plot function now shows a message and returns and empty plot if all effects
  are equal to 0.
* dev.ratio from glmnet is now returned as one of the columns of surrogate_model_explainer.

# localModel 0.3.9

* Explanation models is now fitted to raw predictions, not difference between predictions and model mean.

# localModel 0.3.8

* Fixed bug related to one-colum model matrix in glmnet.

# localModel 0.3.7

* Bug fixed in example.

# localModel 0.3.6

* Improved examples.

# localModel 0.3.5

* Fixed issues with examples.
* Improved documentation.

# localModel 0.3.4

* Fixed a bug related to interpretable numerical inputs (missing as.numeric)
* Added tests for 100% coverage.

# localModel 0.3.3

* Set.seed before fitting cv.glmnet.
* New vignette describing methodology.

# localModel 0.3.2

* Weighting option added.
* Improved documentation.

# localModel 0.3.1

* Improvements to plots 

# localModel 0.3.0

* Multi-model plot option added.
* `stringr` dependency removed.

# localModel 0.2.9

* Same interpretable features are used for every response level.
* `rpart` dependency replaced with `partykit`.

# localModel 0.2.8

* Print method added.

# localModel 0.2.7

* Updates link in README.

# localModel 0.2.6

* Adds help for the package.

# localModel 0.2.5

* Fixes bug related to number of levels when sampling new observation.

# localModel 0.2.4

* Sets seed in tests.

# localModel 0.2.3

* Minor updates to documentation (including paper/paper.md) and unit tests.

# localModel 0.2.2

* Fixed bugs related to column names in case when response in is new_observation.
* Unit tests added.

# localModel 0.2.1

* Vignettes added.

# localModel 0.2.0

* Major changes: interpretable features are created using decision trees-based discretization. 
* Now-unnecessary dependencies dropped.
* Plots are now drawn around intercept.

# localModel 0.1.2

* Fixes bug in observation enconding.
* Improves and simplifies encoding.
* Improves plot.

# localModel 0.1.1

* Improved plot labels.

# localModel 0.1

* Added possibility of coding variables as (baseline / lower or higher) or ( baseline / lower / higher).
* Added option to skip loess smoothing.
* Removed option of setting `family` argument in glmnet.

# localModel 0.0.0.9000

* Added a `NEWS.md` file to track changes to the package.


tests/testthat.R
library(testthat)
library(localModel)
library(DALEX)
library(randomForest)

data('HR')
data('apartments')

m_rf_hr <- randomForest(status ~., data = HR[1:500, ], ntree = 10)
m_rf_ap <- randomForest(m2.price ~., data = apartments[1:500, ], ntree = 10)

hr_explainer <- suppressWarnings(explain(m_rf_hr, HR[1:500, ], verbose = FALSE)) 
ap_explainer <- explain(m_rf_ap, apartments[1:500, ], verbose = FALSE)

local_model_explainer_hr <- individual_surrogate_model(hr_explainer,
                                                       HR[5, -6],
                                                       size = 50,
                                                       seed = 17)
local_model_explainer_ap <- individual_surrogate_model(ap_explainer,
                                                       apartments[5, -1],
                                                       size = 50,
                                                       seed = 17)

test_check("localModel")


tests/testthat/test_main.R
context("Test main functions")


testthat::test_that(
  "Function works with single- and multi-column output",
  testthat::expect_output(individual_surrogate_model(hr_explainer, HR[5, -6], 50),
                          regexp = NA)
  )

testthat::test_that(
  "Reproducibility", {
  testthat::expect_equal(
    individual_surrogate_model(hr_explainer, HR[5, -6], 50, seed = 17),
    individual_surrogate_model(hr_explainer, HR[5, -6], 50, seed = 17)
  )
  testthat::expect_false(
    all(
      individual_surrogate_model(hr_explainer, HR[5, -6], 50) ==
        individual_surrogate_model(hr_explainer, HR[5, -6], 50)
    )
  )
})

testthat::test_that(
  "Non-uniform sampling is okay", {
    testthat::expect_output(
      individual_surrogate_model(hr_explainer, HR[5, -6], 50,
                                 sampling = "non-uniform"),
      regexp = NA
    )
  }
)



tests/testthat/test_generic.R
context("Test generics")

empty_explainer <- local_model_explainer_ap
empty_explainer$estimated <- 0

testthat::test_that("Plots work with different geoms", {
  testthat::expect_is(
    plot(local_model_explainer_ap),
    "gg")
  testthat::expect_is(plot(local_model_explainer_ap, geom = "bar"), "gg")
  testthat::expect_is(plot(local_model_explainer_ap, geom = "arrow"), "gg")
  testthat::expect_message(plot(empty_explainer))    
  })


  testthat::test_that("Print is okay", {
  testthat::expect_output({
    print(local_model_explainer_ap)
  })
})

testthat::test_that("Kernels work fine", {
  testthat::expect_true({
    identity_kernel(HR[2, ], HR[1, ]) == 1
  })
  testthat::expect_false({
    gaussian_kernel(HR[2, 2:5], HR[1, 2:5]) == 1
  })
})



R/interpretable_inputs.R
extract_numerical_feature <- function(rules, true_value) {
  rules_df <- as.data.frame(do.call("rbind", strsplit(rules, " ")),
                            stringsAsFactors = FALSE)
  if(ncol(rules_df) == 7) {
    rules_df[, 3] <- as.numeric(rules_df[, 3])
    rules_df[, 7] <- as.numeric(rules_df[, 7])
    for(row_number in 1:nrow(rules_df)) {
      if(rules_df[row_number, 2] != rules_df[row_number, 6]) {
        lower_limit <- min(rules_df[row_number, 3], rules_df[row_number, 7])
        upper_limit <- max(rules_df[row_number, 3], rules_df[row_number, 7])
        if(lower_limit < true_value & true_value <= upper_limit) {
          label <- paste(c(round(lower_limit, 2),
                           "<", rules_df[1, 1],
                           "<=", round(upper_limit, 2)),
                         sep = " ", collapse = " ")
          lower <- lower_limit
          upper <- upper_limit
          break
        }
      } else {
        if(rules_df[row_number, 2] == "<=") {
          upper_limit <- min(rules_df[row_number, 3], rules_df[row_number, 7])
          if(true_value <= upper_limit) {
            label <- paste(c(rules_df[1, 1], "<=", round(upper_limit, 2)),
                           sep = " ", collapse = " ")
            lower <- -Inf
            upper <- upper_limit
            break
          }
        } else {
          lower_limit <- max(rules_df[row_number, 3], rules_df[row_number, 7])
          if(true_value > lower_limit) {
            label <- paste(c(rules_df[1, 1], ">", round(lower_limit, 2)),
                           sep = " ", collapse = " ")
            lower <- lower_limit
            upper <- Inf
          }
        }
      }
    }
  } else {
    if(true_value <= as.numeric(rules_df[1, 3])) {
      label <- paste(rules_df[1, 1], "<=", round(as.numeric(rules_df[1, 3]), 2),
                     sep = " ", collapse = " ")
      lower <- -Inf
      upper <- as.numeric(rules_df[1, 3])
    } else {
      label <- paste(rules_df[1, 1], ">", round(as.numeric(rules_df[1, 3]), 2),
                     sep = " ", collapse = " ")
      lower <- as.numeric(rules_df[1, 3])
      upper <- Inf
    }
  }

  list(
    "label" = label,
    "interval" = as.numeric(c(lower, upper))
  )
}


extract_categorical_feature <- function(rules, true_value, unique_values,
                                        colname) {
  values <- unique_values[sapply(unique_values, function(value) {
    grepl(value, rules[which(grepl(true_value, rules))]) })]

  list(
    "label" = paste(
      colname, "=",
      paste(values, sep = ", ", collapse = ", ")
    ),
    "values" = as.character(values)
  )
}

#' @author Krystian Igras

`%:::%` <- function (pkg, name) {
  pkg <- as.character(substitute(pkg))
  name <- as.character(substitute(name))
  get(name, envir = asNamespace(pkg), inherits = FALSE)
}

feature_representation <- function(explainer, new_observation, column,
                                   predicted_names, grid_points = 101) {
  is_numerical <- is.numeric(explainer$data[, column])
  n_points <- min(grid_points, length(unique(explainer$data[, column])))

  if(is_numerical) {
    ceteris <- ingredients::ceteris_paribus(
      explainer, new_observation, grid_points = n_points,
      variables = column)[, c(column, "_yhat_", "_label_")]
    if(all(predicted_names == "yhat")) {
      ceteris_curves <- ceteris
      colnames(ceteris_curves)[2] <- "yhat"
    } else {
      ceteris_curves <- data.frame(lapply(predicted_names, function(name) {
        result <- ceteris[ceteris$`_label_` == paste(explainer$label,
                                                     name,
                                                     sep = "."), 2,
                          drop = FALSE]
        colnames(result) <- name
        result
      }))
    }
    ceteris_curves[, column] <- unique(ceteris[, column])
  } else {
    ceteris_curves <- as.data.frame(
      explainer$predict_function(explainer$model,
                                 explainer$data)
    )
    if(ncol(ceteris_curves) == 1) {
      colnames(ceteris_curves) <- "yhat"
    }
    ceteris_curves[, column] <- explainer$data[, column]
  }

  tree_formula <- paste(
    paste(predicted_names, sep = " + ", collapse = " + "),
    column,
    sep = " ~ "
  )

  if(is_numerical) {
    fitted_tree <- partykit::ctree(as.formula(tree_formula),
                                   data = ceteris_curves,
                                   maxdepth = 2)
  } else {
    fitted_tree <- partykit::ctree(as.formula(tree_formula),
                                   data = ceteris_curves,
                                   maxdepth = 1)
  }

  extract_rules <- "partykit" %:::% ".list.rules.party"
  rules <- extract_rules(fitted_tree)
  if(is_numerical) {
    rules <- sapply(rules, function(rule) {
      if(!grepl("&", rule)) {
        paste(rule, rule, sep = " & ", collapse = " & ")
      } else {
        rule
      }
    })
  }

  if(all(rules == " & ") | all(rules == "")) {
    encoded_feature <- as.factor(rep("baseline",
                                     nrow(explainer$data)))
  } else {
    if(is_numerical) {
      interpretable_input <- extract_numerical_feature(rules,
                                                       new_observation[, column])
      encoded_feature <- ifelse(interpretable_input$interval[1] <= explainer$data[, column]
                                & explainer$data[, column] < interpretable_input$interval[2],
                                interpretable_input$label,
                                "baseline")
    } else {
      interpretable_input <- extract_categorical_feature(
        rules,
        new_observation[, column],
        levels(explainer$data[, column]),
        column
      )
      encoded_feature <- ifelse(
        explainer$data[, column] %in% interpretable_input$values,
        interpretable_input$label,
        "baseline"
      )
    }
  }

  factor(encoded_feature,
         levels = c("baseline", setdiff(unique(encoded_feature), "baseline")))
}


R/local_surrogate.R
  calculate_weights <- function(simulated_dataset, new_observation, kernel) {
  for_weights_x <- rbind(simulated_dataset, new_observation)
  for_weights <- cbind(y = 1, for_weights_x)

  model_matrix <- stats::model.matrix(stats::lm(y ~., data = for_weights))
  new_observation_coords <- model_matrix[nrow(model_matrix), ]
  sapply(as.data.frame(t(model_matrix[1:(nrow(model_matrix) - 1), ])),
         function(x) kernel(new_observation_coords, x))
}


#' @importFrom stats as.formula coef model.matrix

single_column_surrogate <- function(x, new_observation,
                                    simulated_data, to_predict,
                                    size, seed = NULL,
                                    weights, sampling = "uniform") {
  predicted_scores <- x$predict_function(x$model, to_predict)

  simulated_data[["y"]] <- 1
  model_mean <- mean(x$predict_function(x$model, x$data))

  model_matrix <- model.matrix(y ~ .,
                               data =  simulated_data)[, -1, drop = FALSE]
  if(ncol(model_matrix) == 1) {
    model_matrix <- cbind(zero = 0, model_matrix)
  }

  if(!is.null(seed)) set.seed(seed)
  fitted_model <- glmnet::cv.glmnet(model_matrix,
                                    predicted_scores,
                                    alpha = 1, weights = weights)
  result <- as.data.frame(as.matrix(coef(fitted_model, lambda = "lambda.min")))
  result$variable <- rownames(result)
  rownames(result) <- NULL
  result <- result[result$variable != "zero", ]
  colnames(result)[1] <- "estimated"
  for(row_number in 2:nrow(result)) {
    result[row_number, "variable"] <- substr(result[row_number, "variable"],
                                             nchar(colnames(simulated_data)[row_number - 1]) + 1,
                                             nchar(result[row_number, "variable"]))
  }
  result <- rbind(
    data.frame(estimated = model_mean,
               variable = "(Model mean)"),
    result
  )
  result$original_variable <- ""
  for(i in 3:nrow(result)) {
    result[i, "original_variable"] <- colnames(new_observation)[
      sapply(colnames(new_observation), function(c) grepl(c, result[i, "variable"]))]
  }
  correct_lambda <- which(fitted_model$glmnet.fit$lambda == fitted_model$lambda.min)
  result$dev_ratio <- fitted_model$glmnet.fit$dev.ratio[correct_lambda]
  result
}


#' LIME-like explanations based on Ceteris Paribus curves
#'
#' This function fits a LIME-type explanation of a single prediction.
#' Interpretable binary features that describe the local impact of features on
#' the prediction are created based on Ceteris Paribus Profiles.
#' Thend, a new dataset of similar observations is created and black box model
#' predictions (scores in case of classification) are calculated for this dataset
#' and LASSO regression model is fitted to them.
#' This way, explanations are simplified and include only the most important features.
#' More details about the methodology can be found in the vignettes.
#'
#' @param x an explainer created with the function DALEX::explain().
#' @param new_observation an observation to be explained. Columns in should correspond to columns in the data argument to x.
#' @param size number of similar observation to be sampled.
#' @param seed If not NULL, seed will be set to this value for reproducibility.
#' @param kernel Kernel function which will be used to weight simulated observations.
#' @param sampling Parameter that controls sampling while creating new observations.
#' @param grid_points Number of points to use while calculating Ceteris Paribus profiles.
#'
#' @return data.frame of class local_surrogate_explainer
#'
#' @export
#'
#' @examples
#' # Example based on apartments data from DALEX package.
#' library(DALEX)
#' library(randomForest)
#' library(localModel)
#' data('apartments')
#' mrf <- randomForest(m2.price ~., data = apartments, ntree = 50)
#' explainer <- explain(model = mrf,
#'                      data = apartments[, -1])
#' model_lok <- individual_surrogate_model(explainer, apartments[5, -1],
#'                                         size = 500, seed = 17)
#' model_lok
#' plot(model_lok)
#'

individual_surrogate_model <- function(x, new_observation, size, seed = NULL,
                                       kernel = identity_kernel,
                                       sampling = "uniform", grid_points = 101) {
  x$data <- x$data[, intersect(colnames(x$data), colnames(new_observation))]
  try_predict <- x$predict_function(x$model, x$data[1:5, ])
  predicted_names <- colnames(try_predict)
  if(is.null(predicted_names))
    predicted_names <- "yhat"

  if(!is.null(seed)) set.seed(seed)
  feature_representations <- lapply(
    colnames(x$data),
    function(column) {
      feature_representation(x,
                             new_observation,
                             column,
                             predicted_names,
                             grid_points)
    }
  )
  encoded_data <- as.data.frame(feature_representations)
  colnames(encoded_data) <- intersect(colnames(new_observation),
                                      colnames(x$data))

  p <- ncol(encoded_data)
  simulated_data <- as.data.frame(
    lapply(encoded_data,
           function(column) {
             as.character(rep(levels(column)[max(1, length(levels(column)))], size))
           }), stringsAsFactors = FALSE)

  probs <- lapply(encoded_data,
                  function(column) {
                    as.data.frame(prop.table(table(column)))$Freq
                  })

  if(!is.null(seed)) set.seed(seed)
  for(row_number in 1:size) {
    n_changes <- sample(1:p, 1)
    change_indices <- sample(1:p, n_changes)
    if(sampling == "uniform") {
      simulated_data[row_number, change_indices] <- "baseline"
    } else {
      for(index in change_indices) {
        simulated_data[row_number, index] <- sample(
          levels(encoded_data[, index]),
          1,
          prob = probs[[index]])
      }
    }
  }

  for(col_number in 1:p) {
    simulated_data[, col_number] <- factor(
      simulated_data[, col_number],
      levels = levels(encoded_data[, col_number])
    )
  }

  n_rows <- nrow(encoded_data)
  if(!is.null(seed)) set.seed(seed)
  to_predict <- data.frame(
    lapply(colnames(simulated_data),
           function(column) {
             how_many_baselines <- sum(simulated_data[, column] == "baseline")
             baseline_indices <- which(encoded_data[, column] == "baseline")
             if(is.numeric(x$data[, column])) {
               ifelse(simulated_data[, column] == "baseline",
                      sample(x$data[baseline_indices, column],
                             how_many_baselines,
                             replace = TRUE),
                      rep(new_observation[, column], size - how_many_baselines)
               )
             } else {
               ifelse(simulated_data[, column] == "baseline",
                      as.character(sample(x$data[baseline_indices, column],
                                          how_many_baselines,
                                          replace = TRUE)),
                      as.character(rep(new_observation[, column],
                                       size - how_many_baselines))
               )
             }
           }))
  colnames(to_predict) <- colnames(simulated_data)
  for(colname in colnames(simulated_data)) {
    if(is.numeric(x$data[, colname])) {
      to_predict[, colname] <- as.numeric(to_predict[, colname])
    } else {
      to_predict[, colname] <- factor(to_predict[, colname],
                                      levels = levels(x$data[, colname]))
    }
  }

  simulated_data <- simulated_data[, sapply(simulated_data,
                                            function(col) length(unique(col)) > 1),
                                   drop = FALSE]
  instance <- data.frame(lapply(simulated_data, function(c) levels(c)[2]))
  weights <- calculate_weights(simulated_data, instance, kernel)

  if(!is.null(ncol(try_predict))) {
    explainer <- lapply(unique(colnames(try_predict)), function(unique_level) {
      internal_explainer <- x
      internal_explainer$predict_function <- function(model, newdata) {
        x$predict_function(model, newdata)[, unique_level]

      }
      result <- single_column_surrogate(internal_explainer,
                                        new_observation,
                                        simulated_data, to_predict,
                                        size, seed,
                                        weights, sampling)
      result[, "response"] <- unique_level
      result[, "predicted_value"] <- internal_explainer$predict_function(
        internal_explainer$model,
        new_observation
      )
      result
    })
  } else {
      explainer <- single_column_surrogate(
        x, new_observation,
        simulated_data, to_predict,
        size, seed, weights, sampling
      )
      explainer[["response"]] <- ""
      explainer[["predicted_value"]] <- x$predict_function(
        x$model,
        new_observation
      )
  }
  if(!is.null(ncol(try_predict))) {
    explainer <- do.call("rbind", explainer)
  }
  attr(explainer, "new_observation") <- new_observation
  explainer$model <- x$label
  class(explainer) <- c("local_surrogate_explainer", class(explainer))
  explainer
}


#' Generic plot function for local surrogate explainers
#'
#' @param x object of class local_surrogate_explainer
#' @param ... other objects of class local_surrogate_explainer.
#' If provided, models will be plotted in rows, response levels in columns.
#' @param geom If "point", lines with points at the end will be plotted,
#' if "bar", bars will be plotted and if "arrow", arrows.
#'
#' @import ggplot2
#'
#' @importFrom stats reorder
#'
#' @export
#'
#' @examples
#' # Example based on apartments data from DALEX package.
#' library(DALEX)
#' library(randomForest)
#' library(localModel)
#' data('apartments')
#' mrf <- randomForest(m2.price ~., data = apartments, ntree = 50)
#' explainer <- explain(model = mrf,
#'                      data = apartments[, -1])
#' model_lok <- individual_surrogate_model(explainer, apartments[5, -1],
#'                                         size = 500, seed = 17)
#' model_lok
#' plot(model_lok)
#'

plot.local_surrogate_explainer <- function(x, ..., geom = "point") {
  variable <- estimated <- intercept <- NULL

  models <- do.call("rbind", c(list(x), list(...)))

  if(all(models$estimated[models$original_variable != ""] == 0)) {
    message("All estimated feature effects are equal to 0.")
    return(ggplot())
  }

  models$labeller <- paste(
    models$model,
    "prediction: ",
    round(models$predicted_value, 2)
  )

  models <- do.call("rbind", by(
    models,
    models$response,
    function(y) {
      y$labeller <- paste(
        unique(y$response),
        paste(unique(y$labeller),
              sep = "\n ", collapse = "\n "),
        sep = "\n ")
      y
    }
  ))

  models$sign <- as.factor(as.character(sign(models$estimated)))
  models <- models[models$variable != "(Intercept)", ]

  models <- do.call("rbind", by(
    models,
    list(models$model, models$response),
    function(y) {
      y$intercept <- y$estimated[y$variable == "(Model mean)"]
      y
    }
  ))

  models <- do.call("rbind", by(
    models,
    list(models$model, models$response),
    function(y) {
      y$estimated = ifelse(
        y$variable == "(Model mean)",
        y$estimated,
        y$estimated + y$intercept
      )
      y
    }
  ))

  models <- do.call("rbind", by(
    models,
    list(models$original_variable),
    function(y) {
      y$all_zero <- all(y$sign == 0)
      y
    }
  ))

  models <- models[!models$all_zero, ]
  models <- models[models$variable != "(Model mean)", ]

  if(geom == "point") {
    final_geom <- geom_pointrange(aes(ymin = intercept, ymax = estimated),
                                  size = 2)
  } else if(geom == "bar") {
    final_geom <- geom_segment(aes(xend = variable, yend = intercept, y = estimated),
                               size = 10, lineend = "butt")
  } else {
    final_geom <- geom_segment(aes(xend = variable, yend = intercept, y = estimated),
                               size = 2,
                               arrow = arrow(length=unit(0.20,"cm"),
                                             ends="first", type = "closed"))
  }

  ggplot(models, aes(x = reorder(variable, -abs(estimated)),
                     y = estimated,
                     color = sign)) +
    theme_bw() +
    geom_hline(aes(yintercept = intercept),
               size = 1)  +
    facet_grid(model~labeller, scales = "free_y") +
    final_geom +
    coord_flip() +
    ylab("Feature influence") +
    xlab("") +
    scale_color_manual(values =  c(`-1` = "#d8b365",
                                   `0` = "#f5f5f5",
                                   `1` = "#5ab4ac")) +
    guides(color = "none")
}


#' Generic print function for local surrogate explainers
#'
#' @param x object of class local_surrogate_explainer
#' @param ... currently ignored
#'
#' @export
#'
#' @importFrom utils head
#'
#' @examples
#' # Example based on apartments data from DALEX package.
#' library(DALEX)
#' library(randomForest)
#' library(localModel)
#' data('apartments')
#' mrf <- randomForest(m2.price ~., data = apartments, ntree = 50)
#' explainer <- explain(model = mrf,
#'                      data = apartments[, -1])
#' model_lok <- individual_surrogate_model(explainer, apartments[5, -1],
#'                                         size = 500, seed = 17)
#' plot(model_lok)
#' model_lok
#'

print.local_surrogate_explainer <- function(x, ...) {
  print(head(as.data.frame(x)))
}
